rust/lance-index/src/reuse.rs

```diff
@@ -0,0 +1,421 @@
+    }
+
+    fn as_vector_index(self: Arc<Self>) -> Result<Arc<dyn crate::vector::VectorIndex>> {
+        Err(Error::NotSupported {
+            source: "FragReuseIndex is not a vector index".into(),
+            location: location!(),
+        })
+    }
+
+    fn statistics(&self) -> Result<serde_json::Value> {
+        let stats = FragReuseStatistics {
+            num_versions: self.details.versions.len(),
+            num_rows: self.details.versions.len(),
+        };
+        serde_json::to_value(stats).map_err(|e| Error::Internal {
+            message: format!("failed to serialize fragment reuse index statistics: {}", e),
+            location: location!(),
+        })
+    }
+
+    async fn prewarm(&self) -> Result<()> {
+        // TODO: can preload the row ID map
+        Ok(())
+    }
+
+    fn index_type(&self) -> IndexType {
+        IndexType::FragmentReuse
+    }
+
+    async fn calculate_included_frags(&self) -> Result<RoaringBitmap> {
+        unimplemented!()
+    }
+}
+
+/// Save a row ID map to a Lance index file.
+pub async fn save_row_id_map(
+    index_store: Arc<dyn IndexStore>,
+    dataset_version: &u64,
+    row_id_map: &HashMap<u64, Option<u64>>,
+) -> Result<String> {
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("old_row_id", DataType::UInt64, false),
+        Field::new("new_row_id", DataType::UInt64, true),
+    ]));
+
+    // need a UUID in case of concurrent compactions writing
+    let file_path = format!("{}_row_id_map_{}.lance", dataset_version, Uuid::new_v4());
+    let mut index_file = index_store
+        .new_index_file(&file_path, schema.clone())
+        .await?;
+
+    let mut old_row_ids = Vec::new();
+    let mut new_row_ids = Vec::new();
+    let mut rows = 0;
+
+    for (old_row_id, new_row_id) in row_id_map.into_iter() {
+        // flush current batch
+        if rows > BATCH_SIZE {
+            let batch = RecordBatch::try_new(
+                schema.clone(),
+                vec![
+                    Arc::new(UInt64Array::from(old_row_ids.clone())),
+                    Arc::new(UInt64Array::from(new_row_ids.clone())),
+                ],
+            )?;
+
+            index_file.write_record_batch(batch).await?;
+
+            old_row_ids.clear();
+            new_row_ids.clear();
+            rows = 0;
+        }
+
+        old_row_ids.push(*old_row_id);
+        new_row_ids.push(*new_row_id);
+        rows += 1;
+    }
+
+    if !old_row_ids.is_empty() {
+        // write remaining
+        let batch = RecordBatch::try_new(
+            schema.clone(),
+            vec![
+                Arc::new(UInt64Array::from(old_row_ids)),
+                Arc::new(UInt64Array::from(new_row_ids)),
+            ],
+        )?;
+
+        index_file.write_record_batch(batch).await?;
+    }
+
+    index_file.finish().await?;
+    Ok(file_path)
+}
+
+pub async fn load_row_id_map(
+    file_path: &str,
+    store: Arc<dyn IndexStore>,
+) -> Result<HashMap<u64, Option<u64>>> {
```

Comment:
If we are going to use this within the conflict resolution code path, I think we should aggressive in limiting IO requests. You can add a unit test that measures the number of read and write IO requests using `StatsHolder`. Here is an example test that does this:

https://github.com/lancedb/lance/blob/01962adcd8d188a1dd1a9daa5ca5ebcc524d9b47/rust/lance/src/index.rs#L1835

Might need to store the file size of the index file in the metadata so that you can skip the additional head request when opening the index file.